flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/metrics/KinesisSourceShardMetrics.java

```diff
@@ -0,0 +1,104 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.kinesis.source.metrics;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.connector.kinesis.source.split.KinesisShardSplit;
+import org.apache.flink.connector.kinesis.source.split.KinesisShardSplitState;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.runtime.metrics.groups.AbstractMetricGroup;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.arns.Arn;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A utility class for handling metrics in {@link
+ * org.apache.flink.connector.kinesis.source.reader.KinesisStreamsSourceReader}.
+ */
+@Internal
+public class KinesisSourceShardMetrics {
+    private static final Logger LOG = LoggerFactory.getLogger(KinesisSourceShardMetrics.class);
+
+    private final Map<String, Long> millisBehindLatestMap = new HashMap<>();
+    private final Map<String, MetricGroup> shardMetricGroupMap = new HashMap<>();
+    private final MetricGroup sourceReaderMetricGroup;
+
+    public KinesisSourceShardMetrics(final MetricGroup metricGroup) {
+        this.sourceReaderMetricGroup = metricGroup;
+    }
+
+    public void registerShardMetricGroup(KinesisShardSplit shard) {
+        LOG.debug("Registering the metric group for shard {}", shard.getShardId());
+        Arn shardArn = Arn.fromString(shard.getStreamArn());
+
+        millisBehindLatestMap.put(shard.getShardId(), -1L);
+        final MetricGroup shardGroup =
+                this.sourceReaderMetricGroup
+                        .addGroup(
+                                MetricConstants.ACCOUNT_ID_METRIC_GROUP, shardArn.accountId().get())
+                        .addGroup(MetricConstants.REGION_METRIC_GROUP, shardArn.region().get())
+                        .addGroup(
+                                MetricConstants.STREAM_METRIC_GROUP, shardArn.resource().resource())
+                        .addGroup(MetricConstants.SHARD_METRIC_GROUP, shard.getShardId());
+
+        shardGroup.gauge(
+                MetricConstants.MILLIS_BEHIND_LATEST_GAUGE,
+                () -> millisBehindLatestMap.get(shard.getShardId()));
+        shardMetricGroupMap.put(shard.getShardId(), shardGroup);
+
+        LOG.debug(
+                "Registered metric identifier {}",
+                shardGroup.getMetricIdentifier(MetricConstants.MILLIS_BEHIND_LATEST_GAUGE));
+    }
+
+    public void updateShardMillisBehindLatest(
+            KinesisShardSplitState shardState, Long millisBehindLatest) {
```

Comment:
We can consider reducing the argument here to minimal necessary (just shardId)

Reply:
It is intended to keep the parameters consistent in this class. I'd like to keep it as is if you don't have concrete concerns.